### PR TITLE
Improve padding logic for updatemenus

### PR DESCRIPTION
**Pull request by rreusser**
See #985 
cc @etpinard 

This PR improves padding logic for updatemenus. It implements `pad: {t, r, b, l}` at the individual-updatemenu level. I didn't apply unit testing, but did beef up the image mock to really ensure that I've tested the permutations.

The general logic:

1. After computing the minimum required size of the container in order to fit the content, it adds the padding onto each side.
2. After expanding the padding, the anchor logic is applied.

The order of operations makes a slight difference, but the general concept is that this applies within the container first, then to the overall layout as expected. I think this is by far the most intuitive and easy way to accomplish this, and meets the need.

Unfortunately it's not that straightforward to document image mocks, and I'd rather not have (literally) sixty mocks to verify all aspects of the positioning, and I think it's best to avoid testing visual alignment in the unit tests.

So on that note, here's the improved mock with the anchor annotated in blue and padding annotated in red:

screen-shot-2016-09-28-at-17 22 49

### CODE REVIEW

**Comment by rreusser** [comment]
Note that I did *not* set a minimum value: 0 for padding. Since we don't share the full generality of the html box model, I feel like negative padding could be a usefully obscure option to have available since we don't have pixel margins to worth with.

**Comment by rreusser** [comment]
Relayout test to be added; battery is at 5% so will be up shortly. Otherwise complete and reviewable.

File: src/components/updatemenus/attributes.js [hunk 1]
```diff
@@ -11,6 +11,7 @@
 var fontAttrs = require('../../plots/font_attributes');
 var colorAttrs = require('../color/attributes');
 var extendFlat = require('../../lib/extend').extendFlat;
+var padAttrs = require('../../plots/pad_attributes');
```
Comment (etpinard) [inline src/components/updatemenus/attributes.js]: That's where `pad_attributes.js` at the moment, but I hate to see see files in `src/plots/` root.

Maybe we add a `src/plots/common/`folder? or `src/components/common/`? or even a `src/common/` for similar shared modules.

File: src/components/updatemenus/draw.js [hunk 2]
```diff
@@ -500,10 +503,10 @@ function findDimenstions(gd, menuOpts) {
...
         x: menuOpts.x,
         y: menuOpts.y,
-        l: menuOpts.totalWidth * ({right: 1, center: 0.5}[xanchor] || 0),
-        r: menuOpts.totalWidth * ({left: 1, center: 0.5}[xanchor] || 0),
-        b: menuOpts.totalHeight * ({top: 1, middle: 0.5}[yanchor] || 0),
-        t: menuOpts.totalHeight * ({bottom: 1, middle: 0.5}[yanchor] || 0)
+        l: paddedWidth * ({right: 1, center: 0.5}[xanchor] || 0),
+        r: paddedWidth * ({left: 1, center: 0.5}[xanchor] || 0),
+        b: paddedHeight * ({top: 1, middle: 0.5}[yanchor] || 0),
+        t: paddedHeight * ({bottom: 1, middle: 0.5}[yanchor] || 0)
```
Comment (etpinard) [inline src/components/updatemenus/draw.js]: Wow. That was easy. I love it!

**Comment by rreusser** [comment]
@etpinard will do. Have run out of steam at the moment, but will aim to have it good to go first thing tomorrow.

File: src/components/updatemenus/attributes.js (hunk 1 above)
Comment (rreusser) [inline src/components/updatemenus/attributes.js]: Agreed. Font attributes would probably go with it? At least internal reorg is easy refactoring, so can try to cut down on `plots/` bit by bit (pun?)

Comment (etpinard) [inline src/components/updatemenus/attributes.js]: > Font attributes would probably go with it?

yep, exactly.



> so can try to cut down on plots/ bit by bit (pun?)

No action required in this PR. I just wanted to mention something I've been thinking about for a while now.

**Comment by rreusser** [comment]
ping @etpinard — added relayout tests that at least confirms a couple basic things about padding getting updated and making some sense.

**Comment by etpinard** [comment]
This is working great. Thanks! :dancer: 

@rreusser merge away

**Pull request merged**